[Merged by Bors] - feat: turn on the `unusedDecidableInType` linter

thorimur opened the pull request:
This PR turns on the `unusedDecidableInType` syntax linter (#31142) after fixing the remaining violations in #31831, #31934, and #32153.

This also imports the linter into Mathlib.Init and adds the linter to the Mathlib standard set. Given that classicality is embedded in Lean, this is probably the behavior downstream repos who use the mathlib standard set want; if they are specifically constructive repos, the linter can always be turned off in the lakefile.

We exempt the entirety of [Mathlib/Data/Fintype/Quotient.lean](https://github.com/leanprover-community/mathlib4/pull/31747/files#diff-9591d2f3828bee189ea954bc33586ce451536bf00792362f7cdca4f5a8551207), [Mathlib/Order/Heyting/Regular.lean](https://github.com/leanprover-community/mathlib4/pull/31747/files#diff-af85d99e5d2f5561fe837f662f179c58f1ebe2a4ed0432a15260199883850c50), [Mathlib/Logic/Encodable/Basic.lean](https://github.com/leanprover-community/mathlib4/pull/31747/files#diff-db01c9a2d4726fe7ca0e6eb18a7cd485b5a9c6b26ba422dfbe501dcebef3cf77) and a declaration in [Mathlib/Computability/Halting.lean](https://github.com/leanprover-community/mathlib4/pull/31747/files#diff-d47323ee8b4b2321052b4be93715f8b2f3f840f76ec441d8bf75c834b8665b32) since constructivity is desired there. We also exempt `dec_em'`, which is a variant of `Decidable.em` but with the disjuncts reversed.

Many prior violations have already been fixed through the use of #10235, hence the small number of violations picked up by the syntax linter.

---

- [x] depends on: #31142 
- [x] depends on: #31831 
- [x] depends on: #31934 
- [x] depends on: #32153
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

thorimur committed:
feat: unused `Decidable*` instance hypotheses linter

thorimur committed:
feat: `withPlural`

thorimur committed:
chore: copyright header

thorimur committed:
chore: fix imports, close section

thorimur committed:
chore: move declarations

thorimur committed:
chore: remove import

thorimur committed:
chore: Mathlib.Init, API

thorimur committed:
chore: more testing

thorimur committed:
chore: fix test

thorimur committed:
chore: change internal `idx`

thorimur committed:
chore: rename to `unusedDecidableInType`

thorimur committed:
chore: remove dependency on `withPlural`

thorimur committed:
chore: make directory dependency exceptions

thorimur committed:
chore: fix docs

thorimur committed:
chore: docstring

thorimur committed:
chore: satisfy unused arguments linter

thorimur committed:
Merge branch 'master' into unusedDecidable-linter

thorimur committed:
chore: module docstring

thorimur committed:
chore: update Mathlib.Tactic.Linter

thorimur committed:
chore: satisfy unusedArguments env linter

thorimur committed:
chore: move to `Mathlib.Lean.Expr.Basic`

thorimur committed:
chore: authorship

thorimur committed:
chore: revert `withPlural` dependency

thorimur committed:
chore: fix unused imports

thorimur committed:
feat: turn on `unusedDecidableInType` linter

thorimur committed:
chore: fix message

thorimur committed:
poke CI

thorimur committed:
chore: fix tests

thorimur committed:
chore: clean up

thorimur committed:
chore: allow constructive exceptions

thorimur committed:
chore: temporary workaround for `withSetOptionIn` breakage

thorimur committed:
feat: improve message formatting

thorimur committed:
chore: fix workaround

thorimur committed:
chore: pluralization

thorimur committed:
chore: make `declFilter` optional

thorimur committed:
chore: don't set option in test

thorimur committed:
chore: update test

thorimur committed:
Merge branch 'master' into unusedDecidable-linter

thorimur committed:
chore: adapt to module system

thorimur committed:
chore: re-update imports after omdule system merge

thorimur committed:
chore: modulize Mathlib.Lean.Environment

thorimur committed:
chore: maybe modulize unusedinstancesintype?

thorimur committed:
chore: modulize in office hours

thorimur committed:
chore: restore Data.Fintype.Quotient exemption that got lost in the merge

thorimur committed:
chore: FiveWheelLike

thorimur commented:
!bench

thorimur committed:
chore: revert stale erroneous binder name change

thorimur committed:
chore: clean up

thorimur committed:
Merge branch 'master' into unusedDecidable-linter

thorimur committed:
chore: clean up after merge

thorimur committed:
chore: clean up

leanprover-radar commented:
Benchmarking bee96eeeb51c4b0b65d2b0690423e1f5dd71dd46 ([status](https://radar.lean-lang.org/repos/mathlib4/commits/bee96eeeb51c4b0b65d2b0690423e1f5dd71dd46)) against b3cf7a02d3df909b93210a065e3be9de016129c5 ([status](https://radar.lean-lang.org/repos/mathlib4/commits/b3cf7a02d3df909b93210a065e3be9de016129c5)).

<sub>React with :eyes: to be notified when the results are in. The command author is always notified.</sub>

thorimur committed:
chore: update test TODO comment

thorimur committed:
chore: `withSetOptionIn` comment link

thorimur committed:
chore: docs about behavior

thorimur committed:
chore: update TODO

thorimur committed:
chore: typo

thorimur committed:
chore: explain filter

thorimur committed:
chore: remove `private`, make option `public`

thorimur committed:
chore: handle `set_option true in`

thorimur committed:
chore: fix matching

thorimur committed:
chore: stray `#check`

thorimur committed:
Update Mathlib/Lean/Expr/Basic.lean

Co-authored-by: damiano <adomani@gmail.com>

thorimur committed:
Update Mathlib/Tactic/Linter/UnusedInstancesInType.lean

Co-authored-by: damiano <adomani@gmail.com>

thorimur committed:
Update Mathlib/Tactic/Linter/UnusedInstancesInType.lean

Co-authored-by: damiano <adomani@gmail.com>

thorimur committed:
chore: update `unusedInstancesMsg` docstring

thorimur committed:
chore: `public` option

thorimur committed:
docs: synchronize docstrings

thorimur committed:
Merge branch 'master' into turn-on-unusedDecidableInType

thorimur committed:
chore: remove manual `set_option` from test file

grunweg commented:
Looks good to me. I'll wait for a second pair of eyes, especially on enabling it by default.
maintainer merge

thorimur committed:
chore: fix Mathlib.Data.Finsupp.Sigma

thorimur committed:
chore: fix Mathlib.Combinatorics.SimpleGraph.Acyclic

thorimur committed:
Merge branch 'fix-more-unused-decidable' into turn-on-unusedDecidableInType

grunweg committed:
chore: import the linter in Mathlib.Init
Needs some downgrading of other imports to avoid build cycles.
TODO: need to fix the lint-style check, probably

grunweg committed:
Enable as part of the mathlib linter set

grunweg committed:
Fix

thorimur committed:
chore: reduce public meta import

thorimur committed:
chore: fix push attr imports

thorimur committed:
chore: exempt `dec_em'`

thorimur committed:
chore: reduce imports further

thorimur committed:
chore: exempt Mathlib.Order.Heyting.Regular?

thorimur committed:
chore: fix imports, remove `meta` where possible

thorimur committed:
chore: fix unusedDecidableInType linter test

thorimur committed:
Merge branch 'master' into turn-on-unusedDecidableInType

thorimur commented:
bors r+
